### ISSUE TITLE
catch22: abstract class vs mixin class

### ISSUE THREAD

**Issue by tomquas** [Iteo/cached#98]
problem: when i generate stubs from the following code, i'll get a dart error:
```
@WithCache()
abstract class ChatAvatarUtils implements _$ChatAvatarUtils {
  factory ChatAvatarUtils() = _ChatAvatarUtils;
    @Cached(ttl: 300) // 5m only cause we only use it for our own avatar
    Future<String> toBase64({required String imageId}) async => '';
}

dart error:
The class 'ChatAvatarUtils' can't be used as a mixin because it's neither a mixin class nor a mixin.
```

however, making it a mixin class and running the generator produces another error:
```
@WithCache()
mixin class ChatAvatarUtils implements _$ChatAvatarUtils {
  factory ChatAvatarUtils() = _ChatAvatarUtils;
    @Cached(ttl: 300) // 5m only cause we only use it for our own avatar
    Future<String> toBase64({required String imageId}) async => '';
}

code generator error:

[ERROR] Class ChatAvatarUtils need to be abstract
package:u2nite/services/image.service.dart:76:13
   ╷
76 │ mixin class ChatAvatarUtils implements _$ChatAvatarUtils {
   │             ^^^^^^^^^^^^^^^
   ╵
```

what can i do?

```
$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.10.0, on macOS 13.3.1 22E772610a darwin-arm64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.78.1)
[✓] Connected device (4 available)
[✓] Network resources

• No issues found!

$ dart --version
Dart SDK version: 3.0.0 (stable) (Thu May 4 01:11:00 2023 -0700) on "macos_arm64"
```

**Comment by Axot017** [comment]
It looks like new dart modifiers changed a little bit how mixins works so code in README is not working on dart >=3.0.0.
As a temporary walkaround you can declare you class as `abstract mixin class` like that:
```dart
@WithCache()
abstract mixin class ChatAvatarUtils implements _$ChatAvatarUtils {
  factory ChatAvatarUtils() = _ChatAvatarUtils;
    @Cached(ttl: 300) // 5m only cause we only use it for our own avatar
    Future<String> toBase64({required String imageId}) async => '';
}
```

We will need to put this walkaround in docs, and investigate if we can change something with generator to make previous version work.

**Comment by PatrykPawlak** [comment]
Hi @tomquas

Sorry for the late response, but we were exploring additional solutions.

We're sticking with the workaround proposed earlier. It's the fix that's doing the trick for Dart >= 3.0.0.

Quick heads up - we've already added the workaround to our docs, so check it out: [Dart 3 changes](https://github.com/Iteo/cached?tab=readme-ov-file#dart-3-changes).

Closing this issue for now. If you or anyone else has more insights or  ace up the sleeve, feel free to drop it here!
